crates/bhyve-api/src/lib.rs

```diff
@@ -193,19 +194,46 @@ impl AsRawFd for VmmFd {
     }
 }
 
-/// Check that bhyve kernel VMM component matches version underlying interfaces
-/// defined in bhyve-api.  Can use a user-provided version (via `version` arg)
-/// or the current one known to bhyve_api.
-pub fn check_version(version: Option<u32>) -> Result<bool> {
-    let ctl = VmmCtlFd::open()?;
-    let current_vers = unsafe {
-        ctl.ioctl(
-            ioctls::VMM_INTERFACE_VERSION,
-            std::ptr::null_mut() as *mut libc::c_void,
-        )
-    }?;
+/// Store a cached copy of the queried API version.  Negative values indicate an
+/// error occurred during query (and hold the corresponding negated `errno`).
+/// A positive value indicates the cached version, and should be less than
+/// `u32::MAX`.  A value of 0 indicates that no query has been performed yet.
+static VERSION_CACHE: AtomicI64 = AtomicI64::new(0);
+
+/// Query the API version from the kernel VMM component on the system.
+///
+/// Caches said version (or any emitted error) for later calls.
+pub fn api_version() -> Result<u32> {
+    fn do_query() -> Result<u32> {
+        let ctl = VmmCtlFd::open()?;
+        let vers = ctl.api_version()?;
+        Ok(vers)
+    }
+
+    if VERSION_CACHE.load(Ordering::Acquire) == 0 {
+        let newval = match do_query() {
+            Ok(x) => x as i64,
+            Err(e) => -(e.raw_os_error().unwrap_or(libc::ENOENT) as i64),
+        };
+        let _ = VERSION_CACHE.compare_exchange(
+            0,
+            newval,
+            Ordering::Acquire,
```

Comment:
nit: should this store have release semantics? I'm out of practice with barrier semantics, so I might have this all wrong, but the Rust docs for `atomic::Ordering` suggest that `Acquire` semantics only provide additional ordering when used with stores that have `Release` semantics. (Or, since we're only dealing with the one memory location, could/should the whole thing just be `Relaxed` everywhere?)

Reply:
Yeah, I think Relaxed is adequate here for now.  I get a little paranoid about memory ordering being too loose, and rust doesn't make cmpxchg any clearer with how it's structured and documented.